python/ray/serve/_private/application_state.py

```diff
@@ -677,3 +788,149 @@ def _save_checkpoint_func(
         )
+
+
+@ray.remote(num_cpus=0, max_calls=1)
+def build_serve_application(
+    import_path: str,
+    app_runtime_env: Dict,
+    config_deployments: List[str],
+    code_version: str,
+    name: str,
+    args: Dict,
+) -> Tuple[List[Dict], Optional[str]]:
+    """Deploy Serve application from a user-provided config.
+
+    Args:
+        import_path: import path to top-level bound deployment.
+        app_runtime_env: application runtime environment. Used to load
+            and import application code
+        config_deployments: list of deployment names specified in config
+            with deployment override options. This is used to check that
+            all deployments specified in the config are valid.
+        code_version: code version inferred from app config. All
+            deployment versions are set to this code version.
+        name: application name. If specified, application will be deployed
+            without removing existing applications.
+        args: Arguments to be passed to the application builder.
+    Returns:
+        Returns None if no error is raised. Otherwise, returns error message.
+    """
+    try:
+        from ray.serve.api import build
+        from ray.serve._private.api import call_app_builder_with_args_if_necessary
+
+        # Import and build the application.
+        app = call_app_builder_with_args_if_necessary(import_attr(import_path), args)
+        app = build(app, name)
+
+        # Check that all deployments specified in config are valid
+        for deployment_name in config_deployments:
+            unique_deployment_name = (
+                (name + DEPLOYMENT_NAME_PREFIX_SEPARATOR) if len(name) else ""
+            ) + deployment_name
+            if unique_deployment_name not in app.deployments:
+                raise KeyError(
+                    f'There is no deployment named "{deployment_name}" in the '
+                    f'application "{name}".'
+                )
+
+        # Set code version and runtime env for each deployment
+        for deployment_name in app.deployments:
+            ray_actor_options = app.deployments[deployment_name].ray_actor_options or {}
+            merged_env = override_runtime_envs_except_env_vars(
+                app_runtime_env, ray_actor_options.get("runtime_env", {})
+            )
+            ray_actor_options.update({"runtime_env": merged_env})
+
+            app.deployments[deployment_name].set_options(
+                version=code_version,
+                ray_actor_options=ray_actor_options,
+                _internal=True,
+            )
+
+        return app.deploy_args_list, None
+    except KeyboardInterrupt:
+        # Error is raised when this task is canceled with ray.cancel(), which
+        # happens when deploy_apps() is called.
+        logger.debug("Existing config deployment request terminated.")
+        return None, None
+    except Exception as e:
+        return None, repr(e)
+
+
+def override_deployment_info(
```

Comment:
please add unit tests for this

Reply:
I've added unit tests to `test_application_state.py`!